accepted/future-releases/primary-constructors/feature-specification.md

```diff
@@ -688,28 +705,47 @@ constructors as well.
 A _declaring constructor_ declaration is a declaration that contains a
 `<declaringConstructorSignature>` with a `<declaringParameterList>`, or a
 declaration that contains a `<declaringConstantConstructorSignature>`, or
-it is a `<primaryConstructorNoConst>` in the header of a class, enum, or
-extension type declaration, together with a declaration in the body that
-contains a `<declaringConstructorSignature>` *(which does not contain a
+it is a `<primaryConstructor>` in the header of a class, enum, or extension
+type declaration, together with a declaration in the body that contains a
+`<declaringConstructorSignature>` *(which does not contain a
 `<declaringParameterList>`, because that's an error)*.
 
 A class or extension type declaration whose class body is `;` is treated as
 a declaration whose body is `{}`.
 
+The grammar is ambiguous with regard to the keyword `factory`.  *For
+example, `factory() => C();` could be a method named `factory` with an
+implicitly inferred return type, or it could be a factory constructor.*
+
+This ambiguity is resolved as follows: When a Dart parser expects to parse
+a `<memberDeclaration>`, and the first token is `factory`, it proceeds to
+parse the following input as a factory constructor.
+
+*Another special exception is introduced with factory constructors in order
+to avoid breaking existing code:*
+
+A factory constructor declaration of the form `factory C(...` where `C`
+is the name of the enclosing class, mixin class, enum, or extension type is
+treated as if `C` had been omitted.
```

Comment:
Would it be more direct to say that it the grammar `factory C(...` is not allowed to parse as a new-style constructor when `C` is the name of the surrounding thing.

Then there is no grammar ambiguity, and the syntax *only* parses as an old-style factory constructor. (And if we ever remove the old-style grammar, we won't still accept `factory C()`, because we forget to remove this special-case. Better to not have it to begin with.)